### PR TITLE
STM32  nor flash controller gets address and size

### PR DESCRIPTION
**Pull request by FRASTM**
the stm32  qspi and ospi flash drivers can retrieve the NOR flash address and size directly from the device tree 
of their `compatible = "st,stm32-qspi-nor";`  or `compatible = "st,stm32-ospi-nor";`
This is done with the <reg> property of the qspi-nor-flash or ospi-nor-flash node as follows:

```
 mx25lm51245: ospi-nor-flash@70000000 {
                 compatible = "st,stm32-ospi-nor";
                reg = <0x70000000 DT_SIZE_M(64)>; /* 512 Mbits */
```

That will make the nor-flash controller driver using the address and size property and will prepare the flash controller for the memory mapped mode (XIP).
That will avoid confusion around the <size> expressed in bits or in byte : here is the reg second param a size in Bytes.

### CODE REVIEW

**Review by erwango — COMMENTED** [review]
Please also update migration guide `doc/releases/migration-guide-3.6.rst`

**Review by Laczen — DISMISSED** [review]
LGTM, it is nice to see the size removed.

File: drivers/flash/flash_stm32_ospi.c [hunk 1]
```diff
@@ -2300,7 +2307,7 @@ static const struct flash_stm32_ospi_config flash_stm32_ospi_cfg = {
 		       .enr = DT_CLOCKS_CELL_BY_NAME(STM32_OSPI_NODE, ospi_mgr, bits)},
 #endif
 	.irq_config = flash_stm32_ospi_irq_config_func,
-	.flash_size = DT_INST_PROP(0, size) / 8U,
+	.flash_size = DT_REG_ADDR_BY_IDX(DT_INST(0, st_stm32_ospi_nor), 1),
```
Comment (erwango) [inline drivers/flash/flash_stm32_ospi.c:2310]: What about using `DT_INST_REG_ADDR_BY_IDX` directly ?

**Review by erwango — CHANGES_REQUESTED** [review]
Otherwise LGTM

**Comment by FRASTM** [comment]
same with DT_INST_REG_ADDR(0) and DT_INST_REG_ADDR_BY_IDX(0, 1)  macro

**Review by erwango — APPROVED** [review]
Thanks

**Review by Laczen — APPROVED** [review]

**Review by de-nordic — APPROVED** [review]

**Comment by erwango** [comment]
CI issue under fix in https://github.com/zephyrproject-rtos/zephyr/pull/68285

**Review by henrikbrixandersen — APPROVED** [review]

**Comment by fabiobaltieri** [comment]
```
/__w/zephyr/zephyr/samples/sensor/lsm6dsl/src/main.c:150:59: error: implicit conversion from 'float' to 'double' when passing argument to function [-Werror=double-promotion]
  150 |                                                           out_ev(&accel_x_out),
      |                                                           ^~~~~~~~~~~~~~~~~~~~
```

CI is sad

**Comment by erwango** [comment]
> CI is sad

CI issue under fix in https://github.com/zephyrproject-rtos/zephyr/pull/68285

**Pull request merged**